### PR TITLE
[feature] tentatively start adding polls support

### PR DESCRIPTION
**Pull request by NyaaaWhatsUpDoc**
# Description
- add support for Question activity in federatingdb
- update internal/ap to handle multiple activity objects
- update federatingdb code to use new ap.ExtractObjects()
- freshen the processing/status.Processor{}.Create() function ready for polls support
- support status update activities, at a very simple level
- start adding in code for some poll support in type converter
- reduces number of allocations during text formatting

this PR is a bit of an odd one of changes here-and-there. it's essentially breaking off most of the remaining poll logic that I can, without introducing the polls themselves, into a separate PR

## Checklist
- [x] I/we have read the [GoToSocial contribution guidelines](https://github.com/superseriousbusiness/gotosocial/blob/main/CONTRIBUTING.md).
- [x] I/we have discussed the proposed changes already, either in an issue on the repository, or in the Matrix chat.
- [x] I/we have performed a self-review of added code.
- [x] I/we have written code that is legible and maintainable by others.
- [x] I/we have commented the added code, particularly in hard-to-understand areas.
- [ ] I/we have made any necessary changes to documentation.
- [ ] I/we have added tests that cover new code.
- [x] I/we have run tests and they pass locally with the changes.
- [x] I/we have run `go fmt ./...` and `golangci-lint run`.

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -14,7 +14,6 @@ run:
 linters:
   # enable some extra linters, see here for the list: https://golangci-lint.run/usage/linters/
   enable:
-    - forcetypeassert
```
Comment (tsmethurst) [inline .golangci.yml:17]: Mm, good idea to remove this. It was getting pretty annoying and not really helping us much.

File: internal/ap/extract.go [hunk 2]
```diff
@@ -35,39 +35,93 @@ import (
...
-	// Extract object vocab.Type.
-	return obj.At(0).GetType()
+	// Accumulate all of the objects into a slice.
+	objTypes := make([]vocab.Type, objProp.Len())
+	for i := 0; i < objProp.Len(); i++ {
+
+		// Get object vocab type at index.
+		objType := objProp.At(i).GetType()
+		if objType == nil {
+			return nil
```
Comment (tsmethurst) [inline internal/ap/extract.go:58]: Hmm, does this mean you're returning nil if even one item in the slice is nil? Might be worth commenting on why.

File: internal/ap/extract.go [hunk 3]
```diff
@@ -35,39 +35,93 @@ import (
...
+	}
+
+	// Accumulate all of the IRIs into a slice.
+	objIRIs := make([]*url.URL, objProp.Len())
+	for i := 0; i < objProp.Len(); i++ {
+
+		// Get iter at index.
+		at := objProp.At(i)
+		if !at.IsIRI() {
+			return nil
```
Comment (tsmethurst) [inline internal/ap/extract.go:88]: Similar to the above: if you have an objects array where the first entry is a Type and the second entry is an IRI (valid ActivityStreams representation), then you're gonna miss out on the IRI here because you've returned early.

File: internal/ap/normalize.go [hunk 4]
```diff
@@ -39,60 +39,36 @@ import (
 // This function is a noop if the type passed in is anything except a Create or Update with a Statusable or Accountable as its Object.
 func NormalizeIncomingActivity(activity pub.Activity, rawJSON map[string]interface{}) {
 	// From the activity extract the data vocab.Type + its "raw" JSON.
-	dataType, rawData, ok := ExtractActivityData(activity, rawJSON)
-	if !ok {
+	dataTypes, rawData, ok := ExtractActivityData(activity, rawJSON)
+	if !ok || len(dataTypes) != len(rawData) {
```
Comment (tsmethurst) [inline internal/ap/normalize.go:43]: In which cases would these two things not be the same? Not asking for changes, just wanna understand.

File: .golangci.yml (hunk 1 above)
Comment (NyaaaWhatsUpDoc) [reply .golangci.yml:17]: this was a commit of pure anger :D

File: internal/federation/federatingdb/accept.go [hunk 5]
```diff
@@ -46,62 +46,22 @@ func (f *federatingDB) Accept(ctx context.Context, accept vocab.ActivityStreamsA
...
-					ReceivingAccount: receivingAccount,
-				})
-
-				return nil
-			}
-		}
+	for _, object := range ap.ExtractObjects(accept) {
+		switch object.GetTypeName() { //nolint:gocritic
+		case ap.ActivityFollow:
+			asFollow := object.(vocab.ActivityStreamsFollow) //nolint:forcetypeassert
```
Comment (tsmethurst) [inline internal/federation/federatingdb/accept.go:52]: might as well do a find and replace for ` //nolint:forcetypeassert` since we're not using that linter as of this PR

File: internal/federation/federatingdb/update.go [hunk 6]
```diff
@@ -103,3 +100,43 @@ func (f *federatingDB) updateAccountable(ctx context.Context, receivingAcct *gts
...
+	idProp := statusable.GetJSONLDId()
+	if idProp == nil || !idProp.IsIRI() {
+		return gtserror.New("invalid id prop")
+	}
+
+	// Get the status URI string for lookups.
+	statusURI := idProp.GetIRI()
+	statusURIStr := statusURI.String()
+
+	// Don't try to update local accounts.
```
Comment (tsmethurst) [inline internal/federation/federatingdb/update.go:115]: local statuses you mean right?

File: internal/processing/status/create.go [hunk 7]
```diff
@@ -39,65 +39,69 @@ import (
...
+	// Get current time.
+	now := time.Now()
+
+	status := &gtsmodel.Status{
+		ID:                       statusID,
+		URI:                      accountURIs.StatusesURI + "/" + statusID,
+		URL:                      accountURIs.StatusesURL + "/" + statusID,
+		CreatedAt:                now,
+		UpdatedAt:                now,
+		Local:                    func() *bool { b := true; return &b }(),
```
Comment (tsmethurst) [inline internal/processing/status/create.go:58]: Don't forget we have that `util.Ptr` function now for such cases as this :)

File: internal/federation/federatingdb/update.go (hunk 6 above)
Comment (NyaaaWhatsUpDoc) [reply internal/federation/federatingdb/update.go:115]: yep :')

File: internal/processing/workers/federate.go [hunk 8]
```diff
@@ -226,6 +226,50 @@ func (f *federate) DeleteStatus(ctx context.Context, status *gtsmodel.Status) er
...
+	}
+
+	// Use ActivityStreams Statusable type as Object of Update.
+	update, err := f.converter.WrapStatusableInUpdate(statusable, false)
+	if err != nil {
+		return gtserror.Newf("error wrapping Statusable in Update: %w", err)
+	}
+
+	// Send the Update activity with Statusable via the Actor's outbox.
+	if _, err := f.FederatingActor().Send(ctx, outboxIRI, update); err != nil {
```
Comment (tsmethurst) [inline internal/processing/workers/federate.go:266]: In future we might want to add some functionality here to send updates to everyone who's liked/boosted the status as well. Not sure exactly how we'd do that, and definitely not saying it needs to be done here. Just a vague sort of reminder to us to do that at some point.

File: internal/ap/normalize.go (hunk 4 above)
Comment (NyaaaWhatsUpDoc) [reply internal/ap/normalize.go:43]: it's more just like an integrity check, i should maybe add a code comment regarding this,

File: internal/ap/extract.go (hunk 3 above)
Comment (NyaaaWhatsUpDoc) [reply internal/ap/extract.go:88]: Ah I was assuming that wouldn't be valid, that an array must all be the same time. Would it perhaps be worth replacing both these functions with a singular one that returns a slice of `TypeOrIRI` ? :thinking:

File: internal/processing/workers/fromfediapi.go [hunk 9]
```diff
@@ -509,6 +513,34 @@ func (p *fediAPI) UpdateAccount(ctx context.Context, fMsg messages.FromFediAPI)
 	return nil
 }
 
+func (p *fediAPI) UpdateStatus(ctx context.Context, fMsg messages.FromFediAPI) error {
```
Comment (tsmethurst) [inline internal/processing/workers/fromfediapi.go:516]: Good stuff!

File: internal/text/plain.go [hunk 10]
```diff
@@ -103,11 +104,11 @@ func (f *Formatter) FromPlainEmojiOnly(
 	statusID string,
 	input string,
 ) *FormatResult {
-	// Initialize standard block parser
-	// that wraps result in <p> tags.
+	// Initialize block parser that
+	// doesn't wrap result in <p> tags.
 	plainTextParser := parser.NewParser(
```
Comment (tsmethurst) [inline internal/text/plain.go:109]: Why the change here you rascal? Doesn't this wildly change how our statuses are serialized?

**Review by tsmethurst — COMMENTED** [review]
Coming along great :) Good work

File: internal/ap/extract.go (hunk 3 above)
Comment (tsmethurst) [reply internal/ap/extract.go:88]: Probably yeah! Mixed arrays are definitely something we need to take account of. They trip me up in JS all the time.

File: internal/text/plain.go (hunk 10 above)
Comment (NyaaaWhatsUpDoc) [reply internal/text/plain.go:109]: This is specifically only in the `FromPlainEmojiOnly()` part, which we only use for pulling emoji out of updated account fields + display names, and formatting status spoiler text. So definitely not places we wanted automatic paragraph HTML tags anywho :p

File: internal/ap/extract.go (hunk 3 above)
Comment (NyaaaWhatsUpDoc) [reply internal/ap/extract.go:88]: argh what a bastard data type!!

File: internal/text/plain.go (hunk 10 above)
Comment (tsmethurst) [reply internal/text/plain.go:109]: Ahh okay okay thanks, I didn't spot that.

File: .golangci.yml (hunk 1 above)
Comment (tsmethurst) [reply .golangci.yml:17]: you and the linter have such a history; enemies to lovers arc when?

File: internal/processing/workers/federate.go (hunk 8 above)
Comment (NyaaaWhatsUpDoc) [reply internal/processing/workers/federate.go:266]: Just thinking, for sending updates to likers / boosters, would we want to send the full updated object or just the updated URI so they can mark as invalidated?

Comment (NyaaaWhatsUpDoc) [reply internal/processing/workers/federate.go:266]: actually wait no AP says we should only use full objects. i'll see if it's a very quick thing i can add in here because if so that would be a useful one to add

Comment (NyaaaWhatsUpDoc) [reply internal/processing/workers/federate.go:266]: i started work on this but stashed the changes as i don't think i can write code i'll be happy with in looking neat + being in the right location if i do it quickly like thiss. :p
